### PR TITLE
Server: add the opensearch product header to threadContext

### LINKED ISSUE #13297: [Feature Request] Add `X-opensearch-product-origin` header to ThreadContext
### Is your feature request related to a problem? Please describe

As someone expand processing request or feature, etc., may want to know it is the context of the OpenSearch dashboard.

### Describe the solution you'd like

Requests from the Opensearch dashboard have the `X-opensearch-product-origin` header. Add this to the ThreadContext header list.

### Related component

Other

### Describe alternatives you've considered

_No response_

### Additional context

https://github.com/opensearch-project/OpenSearch/pull/13296
I created PR for this issue.

### PR DESCRIPTION
**Pull request by llilyy**
### Description

- As someone expand processing request or feature, etc., may want to know it is the context of the OpenSearch dashboard.
- ~~Requests from the Opensearch dashboard have the `X-opensearch-product-origin` header. So I added this to the ThreadContext header list.~~

---

- To distinguish whether the request comes from the opensearch dashboard.
- Requests that come from the Opensearch dashboard have the `x-opensearch-product-origin` header. The value is `opensearch-dashboards`.
- ~~Add the `isDashboardContext` method that returns `boolean` on OpenSearch `ThreadContext`.~~

---

- Add the opensearch product header to threadContext using boolean `TRUE` value.

### Related Issues
Resolves #13297 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
What is the actual use-case where you need to know that the response is coming from OpenSearch? Typically server-side software is expressed with a `Server:` header, but for other custom headers I'd prefer it if we implemented https://github.com/opensearch-project/OpenSearch/issues/1069.

File: server/src/main/java/org/opensearch/rest/RestController.java [hunk 1]
```diff
@@ -330,6 +330,12 @@ private void dispatchRequest(RestRequest request, RestChannel channel, RestHandl
                 // This header is intended for internal use only.
                 client.threadPool().getThreadContext().putHeader(SYSTEM_INDEX_ACCESS_CONTROL_HEADER_KEY, Boolean.FALSE.toString());
             }
+            if (request.header(OPENSEARCH_PRODUCT_ORIGIN_HTTP_HEADER) != null) {
```
Comment (kkmr) [inline server/src/main/java/org/opensearch/rest/RestController.java:333]: Would it be possible to add a simple unit test for this?  Thanks!

**Comment by llilyy** [comment]
@dblock 
Hi, I'm working on a custom OpenSearch plugin that print some logs about search requests of only OpenSearch Dashboard.
Knowing whether the request is for an OpenSearch dashboard may be a general use-case.
The OpenSearch dashboard is already adding the `X-opensearch-product-origin` header to requests. So I don't need to add custom headers to the request.
And I would like to add this header (`X-opensearch-product-origin`) to `ThreadContext` for internal use.

**Comment by dblock** [comment]
@llilyy Thanks, I understand this now, you're just copying the header from the request to the thread context when it's present.

1. Add tests.
2. The code above checks `.... request.header(OPENSEARCH_PRODUCT_ORIGIN_HTTP_HEADER) == null`, reorder the checks to avoid having to lookup twice.
3. What is the side-effect and are there any negative implications of growing the thread context by this header? What if an application sends an extremely large header?

**Comment by jed326** [comment]
> What is the side-effect and are there any negative implications of growing the thread context by this header? What if an application sends an extremely large header?
Does it make sense to allow users to configure this/which headers get copied to thread context this way?
Should this be a boolean or a method on ThreadContext that says the request comes from OpenSearch?

+1 to this, I think allowing the user to copy headers into thread context this way introduces some risk. We could even  specifically check if the `OPENSEARCH_PRODUCT_ORIGIN_HTTP_HEADER` is `opensearch-dashboards` rather than accept all values.

**Comment by llilyy** [comment]
@dblock @jed326 Thanks! I will check and edit codes.

**Review by dblock — CHANGES_REQUESTED** [review]
I think I like this a lot less, I'm sorry :( It makes OpenSearch be aware that the requester is OpenSearch Dashboards, which is a product dependency that doesn't need to exist. WDYT?

**Comment by llilyy** [comment]
@dblock 
I agree. So how about adding a product header with the value `TURE` to `ThreadContext`?

```java
client.threadPool().getThreadContext().putHeader(OPENSEARCH_PRODUCT_ORIGIN_HTTP_HEADER, Boolean.TRUE.toString());
```

**Comment by cwperks** [comment]
FYI The security plugin uses the ThreadContext to keep track of the remote address. It populates a TC header called ([`_opendistro_security_remote_address`](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/auth/BackendRegistry.java#L233-L239)) and takes into account proxied requests by resolving the XFF (X-Forwarded-For) header.

Its technically possible to read that header from another plugin, but that header is not explicitly populated for consumption by downstream plugins. The security plugin does populate [user info](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L214-L226) in the TC for plugins.

**Comment by cwperks** [comment]
I also wanted to mention that the RestController has a feature called `headersToCopy` where it does copy request headers to the TC and you can find the headers list [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/ActionModule.java#L554-L557).

This gets passed to the constructor of the RestController [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/ActionModule.java#L576)

**Comment by dblock** [comment]
The code looks fine. I am not sure what's best, to copy the header (in which case we should probably do what @cwperks suggests if it's already a pattern), or what is here. Maybe I can get some more opinions, @mch2 and @cwperks what would _you_ prefer?

**Comment by cwperks** [comment]
@dblock I'd personally be in favor of using the existing mechanism, but curious to know what maintainers think.

**Comment by cwperks** [comment]
@llilyy Does this comment work for your use-case? https://github.com/opensearch-project/OpenSearch/pull/13296#issuecomment-2079498549

**Comment by llilyy** [comment]
@cwperks Thanks for your sharing. I will check it.

**Comment by llilyy** [comment]
@cwperks 
I confirmed that it works well. thank you! I'll close the pr and issue.

**Pull request closed without merging**